File: leetcode/cpp/arrays_and_hashing/lc_42_trapping_rainwater.cpp
#include <vector>

/**
 * https://leetcode.com/problems/trapping-rain-water/description/
 *
 * Time complexity: O(n) where n = length of height
 * Space complexity: O(n) where n = length of height
 */
class Solution {
public:
    int trap(std::vector<int>& height) {
        // init our max left and right height arrays
        // the max left height array holds the maximum height
        // to the left of each index i in the height array
        std::vector<int> maxLeftHeight(height.size(), 0);
        // the max right height array holds the maximum height
        // to the right of each index i in the height array
        std::vector<int> maxRightHeight(height.size(), 0);

        // first we will populate the max left height array
        // init the first left height to be the same as the height
        // at that location
        maxLeftHeight[0] = height[0];
        // we loop through all max left heights except the first one
        // because water cannot be contained at the left edge (first index)
        for (int i = 1; i < maxLeftHeight.size(); ++i) {
            // populate the max left height as the larger of either
            // the previous max left height or the current left height
            // for each index i
            maxLeftHeight[i] = std::max(height[i-1], maxLeftHeight[i-1]);
        }

        // first we will populate the max right height array
        // init the last right height to be the same as the height
        // at that location
        maxRightHeight[maxRightHeight.size()-1] = height[height.size()-1];
        // we will loop through all max right heights except the last one
        // in reverse because water cannot be contained at the right edge (last index)
        for (int i = height.size()-2; i > 0; --i) {
            // populate the max right height as the larger of either
            // the previous max right height or the current right height
            // for each index i
            maxRightHeight[i] = std::max(height[i+1], maxRightHeight[i+1]);
        }

        // init the totalArea
        int totalArea = 0;
        // loop through all heights except the first and last ones
        // because water cannot be stored at the edges (first and last indices)
        for (int i = 1; i < height.size()-1; ++i) {
            // compute the area as the difference between the smallest max left/right height
            // and the current height at that index
            int area = std::min(maxLeftHeight[i], maxRightHeight[i]) - height[i];

            // if the area is positive, then water can be contained
            // so add it to the total area
            if (area > 0) {
                totalArea += area;
            }
        }

        // finally, return the total area
        return totalArea;
    }
};

File: leetcode/arrays_and_hashing/lc_42_trapping_rain_water.py
from typing import List


'''
https://leetcode.com/problems/trapping-rain-water/description/
'''
class Solution:
    def trap(self, height: List[int]) -> int:
        # init our total area that we will return
        total_area = 0

        # first we will calculate the maxmimum left heights
        # for all the indices, so init our max_left_height array
        max_left_height = [0] * len(height)
        # our base case is that our leftmost max left height
        # is just equal to the leftmost height
        max_left_height[0] = height[0]

        # loop through all height indices starting at 1
        for i in range(1, len(height)):
            # set the max left height to be the 
            # previous max left height or height at
            # the index just before i
            max_left_height[i] = max(max_left_height[i-1], height[i-1])

        # next we will calculate the maxmimum right heights
        # for all the indices, so init our max_right_height array
        max_right_height = [0]*len(height)
        # our base case is that our rightmost max right height
        # is just equal to the rightmost height
        max_right_height[-1] = height[-1]

        # loop through all height indices in reverse from
        # the second last last index
        for i in range(len(height)-2, -1, -1):
            # set the max right height to be the
            # previous max right height or height at
            # the index just after i
            max_right_height[i] = max(max_right_height[i+1], height[i+1])

        # now we can loop through all indices
        # from 1 to second last
        # we don't care about first or last index
        # because water cannot be trapped there
        for i in range(1, len(height)-1):
            # the area at this index is equal to the
            # minimum max height we found - the current height
            # at this index
            area = min(max_left_height[i], max_right_height[i]) - height[i]

            # the area may be negative, so we only
            # add to the total area if it's positive
            if area > 0:
                # add the positive area to the
                # total area
                total_area += area
        
        # finally, we will return the total area
        return total_area
